File: Go/Decirational/token.go
package decirational

import "fmt"

// Token is anything the lexer can emit and the parser can match against.
type Token interface {
	TypeCode() byte
}

// OperandType is the kind of a numeric literal token.
type OperandType byte

const (
	OperandInteger      OperandType = 'i'
	OperandLargeInteger OperandType = 'l'
	OperandRational     OperandType = 'r'
)

func (t OperandType) TypeCode() byte { return byte(t) }
func (t OperandType) String() string {
	switch t {
	case OperandInteger:
		return "INTEGER"
	case OperandLargeInteger:
		return "LARGE_INTEGER"
	case OperandRational:
		return "RATIONAL"
	default:
		return "?"
	}
}

// OperatorKind is a binary arithmetic operator.
type OperatorKind byte

const (
	OpPlus           OperatorKind = '+'
	OpMinus          OperatorKind = '-'
	OpMultiplication OperatorKind = '*'
	OpDivision       OperatorKind = '/'
	// OpIntegerDivision has no single lexable character of its own (its
	// syntax is the two-character "//"), so its underlying byte is just an
	// identity value, never looked up from input; String() below overrides
	// the display text to the real "//" syntax.
	OpIntegerDivision OperatorKind = '\\'
	OpModulo          OperatorKind = '%'
	OpPower           OperatorKind = '^'
)

func (o OperatorKind) TypeCode() byte { return byte(o) }
func (o OperatorKind) String() string {
	if o == OpIntegerDivision {
		return "//"
	}
	return fmt.Sprintf("%c", byte(o))
}

// ParenKind is a parenthesis token.
type ParenKind byte

const (
	LeftParen  ParenKind = '('
	RightParen ParenKind = ')'
)

func (p ParenKind) TypeCode() byte { return byte(p) }
func (p ParenKind) String() string { return fmt.Sprintf("%c", byte(p)) }

// FloorKind is a floor-bracket token.
type FloorKind byte

const (
	LeftFloor  FloorKind = '['
	RightFloor FloorKind = ']'
)

func (f FloorKind) TypeCode() byte { return byte(f) }
func (f FloorKind) String() string { return fmt.Sprintf("%c", byte(f)) }

// AbsoluteKind is the absolute-value bar token. The same bar character opens
// and closes an absolute-value group; which is which is resolved by the
// parser's grammar position.
type AbsoluteKind byte

const AbsoluteBar AbsoluteKind = '|'

func (a AbsoluteKind) TypeCode() byte { return byte(a) }
func (a AbsoluteKind) String() string { return fmt.Sprintf("%c", byte(a)) }

// Operand is a numeric literal token: a value is parsed as a plain machine
// integer when it fits, else as T, else as Rational[T].
type Operand[T CustomInteger[T]] struct {
	Type  OperandType
	Large T           // populated when Type is OperandInteger or OperandLargeInteger
	Rat   Rational[T] // populated when Type is OperandRational
}

func (o Operand[T]) TypeCode() byte { return o.Type.TypeCode() }
func (o Operand[T]) String() string {
	if o.Type == OperandRational {
		return fmt.Sprintf("%s(%s)", o.Type, o.Rat)
	}
	return fmt.Sprintf("%s(%s)", o.Type, o.Large)
}

File: Rust/decirational/src/token.rs
use crate::custom_integer::CustomInteger;
use crate::rational::Rational;
use std::fmt;

/// Every token kind the lexer can produce and the parser consumes.
#[derive(Debug, Clone, PartialEq)]
pub enum Token<T: CustomInteger> {
    /// A literal that fits a 32-bit int.
    Integer(T),
    /// A literal too big for a 32-bit int.
    LargeInteger(T),
    /// A decimal or repeating-decimal literal.
    Rational(Rational<T>),
    Plus,
    Minus,
    Multiply,
    Divide,
    /// "//" - truncating integer division, paired with `Modulo` the same way
    /// `CustomInteger::divide_by`/`modulo` already pair.
    IntegerDivide,
    Modulo,
    Power,
    LeftParen,
    RightParen,
    LeftFloor,
    RightFloor,
    /// The same `|` character opens and closes an absolute-value group;
    /// which is which is resolved by the parser's grammar position.
    AbsoluteBar,
}

impl<T: CustomInteger> fmt::Display for Token<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Token::Integer(v) => write!(f, "INTEGER({})", v),
            Token::LargeInteger(v) => write!(f, "LARGE_INTEGER({})", v),
            Token::Rational(r) => write!(f, "RATIONAL({})", r),
            Token::Plus => write!(f, "+"),
            Token::Minus => write!(f, "-"),
            Token::Multiply => write!(f, "*"),
            Token::Divide => write!(f, "/"),
            Token::IntegerDivide => write!(f, "//"),
            Token::Modulo => write!(f, "%"),
            Token::Power => write!(f, "^"),
            Token::LeftParen => write!(f, "("),
            Token::RightParen => write!(f, ")"),
            Token::LeftFloor => write!(f, "["),
            Token::RightFloor => write!(f, "]"),
            Token::AbsoluteBar => write!(f, "|"),
        }
    }
}
